nixos: make time-sync.target block until initial adjustment with all NTP daemons

thoughtpolice opened the pull request:
###### Motivation for this change

See **[RFC: Accurate time synchronization dependencies via time-sync.target](https://discourse.nixos.org/t/rfc-accurate-time-synchronization-dependencies-via-time-sync-target/1582/1)** on Discourse for more information.

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- [x] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS)
- Built on platform(s)
   - [x] NixOS
- [x] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Assured whether relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

---

thoughtpolice committed:
nixos/boot: distribute systemd-time-wait-sync.service

See:

    https://github.com/systemd/systemd/issues/5097
    https://github.com/systemd/systemd/pull/8494

The basic gist here is that by default, time-sync.target only is
activated upon the *start* of systemd-timesyncd.service, which may
restore a slightly wrong system clock that needs to be adjusted when
initial measurements complete.

The systemd-time-wait-sync service, instead, waits until the kernel
reports a remote synchronized system time and then releases and allows
time-sync.target to trigger. This means that services which require
synchronized clocks can reliably wait on time-sync.target to trigger
before proceeding (if the user is using systemd-timesyncd).

Signed-off-by: Austin Seipp <aseipp@pobox.com>

thoughtpolice committed:
nixos/ntpd: block time-sync.target on ntpwait

Signed-off-by: Austin Seipp <aseipp@pobox.com>

thoughtpolice committed:
nixos/chrony: block time-sync.target on waitsync

Signed-off-by: Austin Seipp <aseipp@pobox.com>

thoughtpolice committed:
nixos/tests: kill chrony hack in cockroachdb test

Now that time-sync.target correctly triggers on the NTP daemon
performing its first major adjustment after boot(rather than simply
starting the daemon process), we can remove this hack in the test that
required specific knowledge of Chrony as the NTP daemon, to be injected
into CockroachDB's startup logic.

(Note that we still continue to use Chrony anyway, since it's the only
NTP daemon that can synchronize against PTP clocks on Linux. However,
the real result is that CockroachDB will work properly now for users,
regardless of their chosen daemon.)

Signed-off-by: Austin Seipp <aseipp@pobox.com>

jccleaver commented:
The basic concept of "do large scale time sync at boot before engaging ntp for skew" makes sense and is how things work in (pre-systemd) EL6 still. If you enable both ntpdate (gross adjustment, one-time hit) and ntp (ntpd), the former occurs first via basic sysvinit ordering.

Mic92 commented:
I would prefer if this would not delay my laptop/desktop bootup time, where I don't care about clock skew at all similar to: https://github.com/NixOS/nixpkgs/pull/44524#issuecomment-431648123

Mic92 commented:
But I guess your plan was to make this some kind of optional target clustered software can depend on.

thoughtpolice commented:
> I would prefer if this would not delay my laptop/desktop bootup time, where I don't care about clock skew at all similar to: #44524 (comment)

`time-sync.target` is only depended upon by few critical services (Ceph and Cockroach as of this writing) and `multi-user.target` isn't one of them -- though bringing *in* Ceph on your desktop or Laptop will transitively do so. If your system requires software that must have accurate time-synchronization, then it's unavoidable: it's a correctness concern.

But realistically most desktop software will be fine for the first synchronization, and this won't get pulled into `multi-user.target`. It could still happen somehow, though-- for example, a [Roughtime](https://blog.cloudflare.com/roughtime) service could be used in conjunction with NTP on a piece of desktop software, but you probably aren't going to see this today or anytime soon. (Just making the possibility clear.)

Also, realistically speaking, NTP is *extremely important* as an undocumented dependency for a substantial amount of software, clustering databases aside -- an extreme amount of network-oriented software at all implicitly relies on clock synchronization. But adding `time-wait.sync` to everything isn't practical, either, and I have no intentions of changing this.

> But I guess your plan was to make this some kind of optional target clustered software can depend on.

I would argue it should be non-optional in the cases of CockroachDB or Ceph, i.e. they should be `Requires=/Before=` -- because proper synchronization is critical to prevent node failure or catastrophic error. In these cases the daemon either will fail (for some reason) or the waiting process will timeout (to be done, this patch doesn't just yet) resulting in the service failing too, which is the appropriate course of action, I would argue.

Other software, if appropriate, can relax the constraint by moving to `Wants=/Before=` and even dropping `Before=` (to allow concurrent execution) if time is less important but still necessary.

thoughtpolice committed:
nixos/chrony: add flags to control initial boot time adjustment

By default, 'chronyc waitsync' waits forever which is clearly
unacceptable. Even then, it also takes a bit *too long* to do initial
adjustment checking; in particular it has a surprisingly long (~10s)
testing interval, which can delay things significantly.

By default, check 12 times with a 5 second interval before giving up --
yielding a 60 second initial boot-time adjustment timeout.

Signed-off-by: Austin Seipp <aseipp@pobox.com>

thoughtpolice committed:
nixos/chrony: add flag to skip initial boot time adjustment

In certain scenarios running Chrony, chronyc waitsync will never return
or will erroneously timeout while performing the initial boot
adjustment. In cases like this, and maybe some other strange, advanced
scenarios, it may be useful to not wait on the initial boot time
adjustment.

The most apparent reason for this that I have encountered is that
'chronyc waitsync' apparently does not understand 'local stratum'
directives, which are used to run local NTP daemons that are
disconnected from any servers/reference clocks (e.g. in an offline
network). In this case, 'waitsync' apparently never returns even though
the clock is synchronized vacuously. But even if this was handled
'correctly' -- in these cases, there is no need to wait on initial
synchronization anyway.

The most useful application of this feature currently is for NixOS
tests: this allows us to run NixOS tests that run NTP daemon nodes
(disconnected from any network/physical clock), which other nodes can
synchronize to.

(It's unclear to me if 'waitsync' not immediately returning in these
cases of offline 'local stratum' servers is a bug or intended behavior.)

Signed-off-by: Austin Seipp <aseipp@pobox.com>

thoughtpolice commented:
@GrahamcOfBorg nixosTests.cockroachdb

thoughtpolice commented:
Note: I am trying to get this into 19.03 (since it is basically essential for reliable use of Ceph and CockroachDB, two modules we will be shipping) and have made a number of bugfixes to the Chrony module, as well as the CockroachDB tests, which I am using to battletest the NTP daemons.

thoughtpolice commented:
@GrahamcOfBorg build nixosTests.cockroachdb

mmahut commented:
Are there any updates on this pull request, please?

thoughtpolice commented:
@mmahut I apologize, I need to split out a few of these commits I think which can go upstream immediately. Then I will return and fix the rest.

thoughtpolice committed:
nixos/tests: make cockroachdb test more flexible

This greatly enhances the flexibility of the CockroachDB test and takes
care of a number of issues with the previous one:

  - Workload and workload duration are now parameterizable.
  - The number of nodes, and their memory requirements, is also
    parameterizable.
  - The Cockroach nodes themselves now synchronize to a fourth node
    which exposes a chrony server. This can be synchronized in some
    arbtirary way.
  - As a result of the prior point, the CRDB nodes can have any
    NTP daemon they like configured.
  - As a result of the second point (again), we don't have to
    synchronize the chrony node to the PTP clock if we don't want
    to. This should make the tests more portable, particularly
    for aarch64, as a skewed clock can still serve as a source of
    truth for the other nodes.
  - Now enabled as a real NixOS test!

Signed-off-by: Austin Seipp <aseipp@pobox.com>

thoughtpolice commented:
This branch has now been rebased on top of master and should build cleanly; the CockroachDB test has been rewritten in this PR, but once the details are sorted out, I will likely move that to *another* PR to be merged after this.

Mic92 commented:
> > I would prefer if this would not delay my laptop/desktop bootup time, where I don't care about clock skew at all similar to: #44524 (comment)
> 
> `time-sync.target` is only depended upon by few critical services (Ceph and Cockroach as of this writing) and `multi-user.target` isn't one of them -- though bringing _in_ Ceph on your desktop or Laptop will transitively do so. If your system requires software that must have accurate time-synchronization, then it's unavoidable: it's a correctness concern.
> 

Sounds good to me then.

Mic92 commented:
This one is still relevant.

nixos-discourse commented:
This pull request has been mentioned on **NixOS Discourse**. There might be relevant details there:

https://discourse.nixos.org/t/systemd-wait-for-timesync/15808/1

NickCao commented:
Still relevant, at least we should add `systemd-time-wait-sync.service` to `systemd.additionalUpstreamSystemUnits`, if not the other changes.
